File: cultraview/VirtualKey/floatwindowlib/src/main/java/com/protruly/floatwindowlib/ui/SelectAppDialog.java
package com.protruly.floatwindowlib.ui;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.AdapterView;
import android.widget.FrameLayout;
import android.widget.GridView;

import com.protruly.floatwindowlib.R;
import com.protruly.floatwindowlib.been.AppInfo;
import com.protruly.floatwindowlib.utils.ApkInfoUtils;
import com.yinghe.whiteboardlib.utils.CommConst;
import com.yinghe.whiteboardlib.utils.SPUtil;
import com.zhy.adapter.abslistview.CommonAdapter;
import com.zhy.adapter.abslistview.ViewHolder;
import com.zhy.http.okhttp.utils.L;

import java.util.List;

/**
 * Created by gaoyixiang on 2019/5/20.
 */

public class SelectAppDialog extends FrameLayout {

    private ApkInfoUtils apkInfoUtils;
    private List<AppInfo> dataList;
    public Callback mCallback;
    private Context mContext;
    private GridView gridView;
    public SelectAppDialog(@NonNull Context context) {
        super(context);
        this.mContext=context;
        init();
    }

    public SelectAppDialog(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        this.mContext=context;
        init();
    }

    public SelectAppDialog(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        this.mContext=context;
        init();
    }
    interface Callback {
        void selected(AppInfo appInfo);
    }

    private AdapterView.OnItemClickListener mOnItemClickListener = new AdapterView.OnItemClickListener() {
        @Override
        public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
            L.e("position"+position);
            if(position < dataList.size()){
                AppInfo appInfo = dataList.get(position);
                if (appInfo != null){
                    if(mCallback != null){
                        mCallback.selected(appInfo);
                        SPUtil.saveData(getContext(), CommConst.USERED_PACKAGE_NAME, appInfo.getPackName());
                    }
                }
            }
        }
    };

    private void init(){
        apkInfoUtils = new ApkInfoUtils();
        LayoutInflater.from(mContext).inflate(R.layout.dialog_apps, this);
        initView();
    }

    private void initView() {
        dataList = apkInfoUtils.scanInstallApp(mContext);
        int item_app = R.layout.item_app;
        CommonAdapter<AppInfo> commonAdapter=new CommonAdapter<AppInfo>(mContext,item_app,dataList) {
            @Override
            protected void convert(ViewHolder viewHolder, AppInfo appInfo, int position) {
                viewHolder.setText(R.id.text,appInfo.appName);
                viewHolder.setImageDrawable(R.id.image,appInfo.appIcon);
            }
        };

        gridView = findViewById(R.id.signal_grid_view);
        gridView.setAdapter(commonAdapter);
        gridView.setOnItemClickListener((adapterView, view, position, l) -> {
            L.e("position"+position);
            if(position < dataList.size()){
                AppInfo appInfo = dataList.get(position);
                if (appInfo != null){
                    if(mCallback != null){
                        mCallback.selected(appInfo);
                        SPUtil.saveData(getContext(), CommConst.USERED_PACKAGE_NAME, appInfo.getPackName());
                    }
                }
            }
        });
    }

    public void setCallBack(Callback mCallBack){
        this.mCallback=mCallBack;
    }
}

File: cultraview/easytouch/app/src/main/java/com/ctv/easytouch/ui/SelectAppDialog.kt
package com.ctv.easytouch.ui

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.AdapterView
import android.widget.FrameLayout
import android.widget.GridView
import com.ctv.easytouch.R
import com.ctv.easytouch.been.AppInfo
import com.ctv.easytouch.utils.ApkInfoUtils
import com.zhy.adapter.abslistview.CommonAdapter
import com.zhy.adapter.abslistview.ViewHolder

class SelectAppDialog @JvmOverloads constructor(private val mContext: Context, attrs: AttributeSet? = null) :
    FrameLayout(mContext, attrs) {
    private var apkInfoUtils: ApkInfoUtils? = null
    private var dataList: List<AppInfo>? = null

    private var mCallback: Callback? = null

    /**
     * item监听器
     */
    private val mOnItemClickListener = AdapterView.OnItemClickListener{ parent, view, position, id ->
        if (position < dataList!!.size) {
            val appInfo = dataList!![position]
            if (appInfo != null) {
                if (mCallback != null) {
                    mCallback!!.selected(appInfo)
                }
            }
        }
    }

    interface Callback {
        fun selected(appInfo: AppInfo)
    }

    init {

        init()
    }

    private fun init() {
        apkInfoUtils = ApkInfoUtils()

        LayoutInflater.from(mContext).inflate(R.layout.dialog_signal, this)
        initView()
    }

    /**
     * 初始化UI
     */
    private fun initView() {

        // 获得数据
        dataList = apkInfoUtils!!.scanInstallApp(mContext)

        // 设置Adapter
        val gridMenuID = R.layout.item_app
        val commonAdapter = object : CommonAdapter<AppInfo>(
            mContext,
            gridMenuID, dataList
        ) {
            override fun convert(holder: ViewHolder, appInfo: AppInfo, position: Int) {
                holder.setText(R.id.text, appInfo.appName)
                holder.setImageDrawable(R.id.image, appInfo.appIcon)
            }
        }

        // 初始化
        val gridView = findViewById<View>(R.id.signal_grid_view) as GridView
        gridView.adapter = commonAdapter
        gridView.onItemClickListener = mOnItemClickListener
    }

    fun setCallback(mCallback: Callback) {
        this.mCallback = mCallback
    }
}
